File: Concurrency/1226_The_Dining_Philosophers/Python/Solution.py
import threading

class DiningPhilosophers:
    def __init__(self):
        # number of philosophers
        self.numOfPhilosophers = 5
        # set up mutex locks
        self.forks = []
        for i in range(self.numOfPhilosophers):
            self.forks.append(threading.Lock())

    # call the functions directly to execute, for example, eat()
    def wantsToEat(
        self,
        philosopher: int,
        pickLeftFork: 'Callable[[], None]',
        pickRightFork: 'Callable[[], None]',
        eat: 'Callable[[], None]',
        putLeftFork: 'Callable[[], None]',
        putRightFork: 'Callable[[], None]',
    ) -> None:
        # left and right forks
        leftForkIdx = (philosopher + 1) % self.numOfPhilosophers
        rightForkIdx = philosopher

        # divide order of fork acquiring to prevent deadlocks
        if philosopher % 2 == 0:
            # lock left fork and pick it up
            self.forks[leftForkIdx].acquire()
            pickLeftFork()
            # lock right fork and pick it up
            self.forks[rightForkIdx].acquire()
            pickRightFork()
        else:
            # lock right fork and pick it up
            self.forks[rightForkIdx].acquire()
            pickRightFork()
            # lock left fork and pick it up
            self.forks[leftForkIdx].acquire()
            pickLeftFork()

        # eat spaghetti
        eat()

        # unlock left fork and put it down
        putLeftFork()
        self.forks[leftForkIdx].release()
        # unlock right fork and put it down
        putRightFork()
        self.forks[rightForkIdx].release()

File: Concurrency/1226_The_Dining_Philosophers/Go/Solution.go
package main

import (
	"fmt"
	"sync"
)

// function type for passing function to methods
type PrinterFunc func()

type DiningPhilosophers struct {
	// number of philosophers
	NumOfPhilosophers int
	// mutex locks
	Forks []sync.Mutex
}

// initialize Dining Philosophers solution struct
func (dp *DiningPhilosophers) Init(numOfPhilosophers int) {
	// set up number of philosophers
	dp.NumOfPhilosophers = numOfPhilosophers
	// initialize mutex locks
	dp.Forks = make([]sync.Mutex, dp.NumOfPhilosophers)
}

// instruct philosopher to eat
func (dp *DiningPhilosophers) WantsToEat(
	philosopher int,
	pickLeftFork PrinterFunc,
	pickRightFork PrinterFunc,
	eat PrinterFunc,
	putLeftFork PrinterFunc,
	putRightFork PrinterFunc) {

	// left and right forks
	leftForkIdx := (philosopher + 1) % dp.NumOfPhilosophers
	rightForkIdx := philosopher

	// divide order of fork locking to prevent deadlocks
	if philosopher%2 == 0 {
		// lock left fork and pick it up
		dp.Forks[leftForkIdx].Lock()
		pickLeftFork()
		// lock right fork and pick it up
		dp.Forks[rightForkIdx].Lock()
		pickRightFork()
	} else {
		// lock right fork and pick it up
		dp.Forks[rightForkIdx].Lock()
		pickRightFork()
		// lock left fork and pick it up
		dp.Forks[leftForkIdx].Lock()
		pickLeftFork()
	}

	// eat spaghetti
	eat()

	// unlock left fork and put it down
	putLeftFork()
	dp.Forks[leftForkIdx].Unlock()
	// unlock right fork and put it down
	putRightFork()
	dp.Forks[rightForkIdx].Unlock()
}

func main() {
	// number of times each philosopher wants to eat
	n := 5
	// number of philosophers
	numOfPhilosophers := 5
	// wait group for goroutine synchronization
	wg := sync.WaitGroup{}
	// add number of times wantsToEat() will be called to wait group
	wg.Add(n * numOfPhilosophers)

	// create and initialize Dining Philosophers solution struct
	dp := DiningPhilosophers{}
	dp.Init(numOfPhilosophers)

	for i := 0; i < n; i++ {
		for id := 0; id < numOfPhilosophers; id++ {
			// go routine for philosopher to eat
			go func(philosopher int) {
				dp.WantsToEat(
					philosopher,
					func() { fmt.Println(philosopher, 1, 1) },
					func() { fmt.Println(philosopher, 2, 1) },
					func() { fmt.Println(philosopher, 0, 3) },
					func() { fmt.Println(philosopher, 1, 2) },
					func() { fmt.Println(philosopher, 2, 2) },
				)
				// signal wait group termination
				wg.Done()
			}(id)
		}
	}

	// wait for all waitgroups to be done
	wg.Wait()
}
